### PR TITLE
WIP - Changes to support comparing histograms

### PR DESCRIPTION
**Pull request by Perryo**
@jakepruitt 
Can you check this out?
We need to update the block with the quantized values for this to work as we build the hist off the block. Lets look into that.

Basically I created a dictionary of dictionaries `frame_block_dict` whose key is the frame number. This dictionary contained in `frame_block_dict` is the dictionary of the block `block_hist_dict` whose key is a tuple `(x,y)` the block coordinate and which contains the histogram for that block. 

Then in the compare function I create the histograms for the blocks of the target frame, and using the dictionary data structure loop through and compare all the histograms of the frames.

This still needs to be adapted for Gabriel's section.

### CODE REVIEW

File: Project 3/block_quantize.py [hunk 1]
```diff
@@ -69,9 +73,13 @@ def quantize(frame_data, n):
         for block_x in range(0, len(frame), 8):
             for block_y in range(0, len(frame[block_x]), 8):
                 block = frame[block_x:block_x+8, block_y:block_y+8]
-                result.extend(quantize_block(block, n, frame_num, block_x, block_y))
+                block_dict,block_hist = quantize_block(block, n, frame_num, block_x, block_y)
+                block_hist_dict[block_x,block_y] = block_hist
+                result.extend(block_dict)
+
+            frame_block_dict[frame_num] = block_hist_dict
```
Comment (Perryo) [inline Project 3/block_quantize.py:93]: This is where we add the histogram to the dictionary for the histograms of the blocks where the coordinates are the key. We then add that dictionary to the frame dictionary where the frame number is the key to access the dictionaries of the regions.

File: Project 3/match_frame.py [hunk 2]
```diff
@@ -72,6 +72,47 @@ def show_ten_closest(frame_data, feature_summary, frame_num, description):
...
+    top_ten_frames = list((x[0] for x in top_ten_frames))    #just need to the frame number, not the diff so we strip that out
+    for i in range(0,10):
+        # For each of those frame numbers, display the image in a window with the number
+        index = top_ten_frames[i]
+        rgb_target = cv2.cvtColor(frame_data[index].astype(np.uint8), cv2.COLOR_GRAY2BGR)
+        cv2.imshow(description + ' ' + str(i), rgb_target)
+        cv2.waitKey(0)
+
+    cv2.destroyAllWindows()
+
```
Comment (Perryo) [inline Project 3/match_frame.py:110]: This starts by building the block histograms for the target frame, then accesses the structures made above to compare the target frame histograms against all the frames histograms.

**Comment by Perryo** [comment]
@jakepruitt @GCrispino @SreeniASU @ApertureTS 

Currently the cv2.compareHist function is returning the same total values for all the histograms making only the last frames show up. Can someone take a look at this? We may need to adopt a different compare method.

**Comment by jakepruitt** [comment]
Looks good from what I can read so far, going to pull down and give it a run.

**Comment by jakepruitt** [comment]
@Perryo Getting this error right now:

`cv2.error: /tmp/opencv20150906-42178-3d0iam/opencv-2.4.12/modules/imgproc/src/histogram.cpp:192: error: (-215) ranges[i] && ranges[i][0] < ranges[i][1] in function histPrepareImages`

In the call for calculating the histogram. Investigating now.

**Comment by jakepruitt** [comment]
@Perryo the reason for that ^ error was for a frame block that had all of the same value. I'm not sure how to handle this at the moment, I imagine it would probably be a histogram with `[64,0,0,0...0]` for however long the number of bins is...

Inspecting the histograms created from the `calcHist` function in openCV, it looks like it's just an array with length `n` where `n` is the number of bins provided by the user. Each value in that array is the frequency count in that bin, so the arrays would look like this:

```
[12, 31, 0, 2, 9, 10]
```

Note that the counts always add up to 64 - the number of pixels in a block.

Now, we already get this information from the existing implementation - in fact we have even more information since we know the value these counts relate to:

```
{
30: 12,
45: 31,
60: 0,
75: 2,
90: 9,
105: 10
}
```

^ We can probably save computation time and complexity by taking our existing data structure created by `quantize_block` and creating the histogram that opencv expects by taking the counts and creating an ordered array from it.

In terms of the `compareHist` function, why did you go with Chi squared? How will we determine the best way to compare histograms? [Here's a stackoverflow answer about the subject](http://stackoverflow.com/questions/6499491/comparing-two-histograms), and [here's the source code for OpenCV's compareHist function](https://github.com/Itseez/opencv/blob/debe99f1a4cc0ea859b91996613934e361b544b6/modules/imgproc/src/histogram.cpp#L2333-L2538). I'll have to do some reading on it myself.

**Comment by Perryo** [comment]
@jakepruitt 

I just made some changes and it should all be working now. The scope of the block_hist_dict was too broad and led to it being the same for every frame, I fixed that. As far as chi-squared, I just wanted a lower diff, but correlation would work better from what I've seen, but larger values = higher correlation so we would need to reverse the result array. Still very simple. I can go ahead and implement that. 

http://www.pyimagesearch.com/2014/07/14/3-ways-compare-histograms-using-opencv-python/

**Comment by Perryo** [comment]
@jakepruitt 

Forgot to mention, as far as complexity is concerned the comparison of frames takes _maybe_ a second on my machine, I dont see it worth optimizing at this point given that the implementation is working (at least from what I can see, run video R2 with the target frame as 160 and you will see what I mean)

**Comment by Perryo** [comment]
@jakepruitt  @GCrispino 
Is it okay to merge this? Have you guys checked it out? Its working good from what I can see

**Comment by GCrispino** [comment]
@Perryo @jakepruitt
I have not checked that out yet, I apologize. I will do that later today. But if Jake says it is all right, I think we are good, because you guys worked the most on this part.

**Comment by jakepruitt** [comment]
@Perryo added fix for when `max_value == min_value`, which was causing a bug. A few notes:
1. The results coming back for block-level quantization are not very good. When I run `python match_frame.py 150 5 5 ../test/project1/R2.mp4` on the 150th frame in the video, which has a boy's face in the center of the frame, none of the other frames created by quantization have the boy's face - most of them are just frames without anything in them.
2. As for potential performance improvements, I'm not referring to frame _comparison_ but histogram _calculation_ being slower than needed. The operation done by `calcHist` is already done by the current code, we just need to add some code to transform the current outputted dictionary into the histogram format. This is not a priority right now, but I do think it might be worth it to change calculation time of a small video from ~ 1minute to ~ 5 seconds.

I'm going to work on a branch that logs the scores of the best frames, which I think is the only thing needed after this. I would say merge this branch and we can refine results and performance if needed.

**Comment by jakepruitt** [comment]
Also changed task 1d to use histogram diffing. I'm going to go ahead and merge this as is and keep momentum going. We can add more adjustments soon.

**Comment by Perryo** [comment]
Did you pull the latest? It's matching perfectly from my last build.

**Pull request merged**